test/utils/readRowsImpl.ts

```diff
@@ -154,80 +170,216 @@ function isKeyInRowSet(
   return true;
 }
 
-// Returns an implementation of the server streaming ReadRows call that would return
-// monotonically increasing zero padded rows in the range [keyFrom, keyTo).
-// The returned implementation can be passed to gRPC server.
-export function readRowsImpl(
-  keyFrom: number,
-  keyTo: number,
-  errorAfterChunkNo?: number
-): (
-  stream: ServerWritableStream<
-    protos.google.bigtable.v2.IReadRowsRequest,
-    protos.google.bigtable.v2.IReadRowsResponse
-  >
-) => Promise<void> {
-  return async (
-    stream: ServerWritableStream<
-      protos.google.bigtable.v2.IReadRowsRequest,
-      protos.google.bigtable.v2.IReadRowsResponse
-    >
-  ): Promise<void> => {
-    prettyPrintRequest(stream.request);
+/** Gets the key value for the given property specified in the request.
+ * @param request The request object to get the key value from.
+ * @param property The property from the request to get the value from.
+ * @returns {string | undefined} The key value from the request.
+ */
+function getKeyValue(
+  request: protos.google.bigtable.v2.IReadRowsRequest,
+  property: keyof IRowRange
+) {
+  if (
+    request?.rows?.rowRanges &&
+    request?.rows?.rowRanges[0] &&
+    request?.rows?.rowRanges[0][property]?.toString()
+  ) {
+    return request?.rows?.rowRanges[0][property]?.toString();
+  }
+  return undefined;
+}
+
+/** Gets the key from the request object.
+ * @param stream The stream object to get the key from.
+ * @param keySelectionParameters The parameters for selecting the key.
+ * @returns {number} The selected key for generating chunks
+ */
+function getSelectedKey(
+  request: protos.google.bigtable.v2.IReadRowsRequest,
+  keySelectionParameters: {
+    keyOpenProperty: keyof IRowRange;
+    keyClosedProperty: keyof IRowRange;
+    defaultKey?: number;
+  }
+) {
+  const keyRequestOpen = getKeyValue(
+    request,
+    keySelectionParameters.keyOpenProperty
+  );
+  const keyRequestClosed = getKeyValue(
+    request,
+    keySelectionParameters.keyClosedProperty
+  );
+  const defaultKey = keySelectionParameters.defaultKey;
+  return defaultKey !== undefined
+    ? defaultKey
+    : keyRequestClosed
+      ? parseInt(keyRequestClosed as string)
+      : parseInt(keyRequestOpen as string) + 1;
```

Comment:
These combined ternary expressions are hard to read. Is this saying if the defaultKey is not undefined, then use default key, else use keyRequestClosed, and if that is undefined use it as opened, or else as closed? I think this might achieve the same thing and be easier to read, or if not maybe you can rewrite it in a way that preserves the logic:
```
return defaultKey || keyRequestClosed ? parseInt(keyRequestClosed as string) : parseInt(keyRequestOpen as string) + 1;
```

Reply:
We can't do it quite like this because when defaultKey is 0 it is falsy so it doesn't get used, but we want it to get used when it is 0. What I did instead was did an equal check instead of a not equal check.